cli/flox-rust-sdk/src/providers/buildenv.rs

```diff
@@ -158,20 +170,71 @@ impl BuildEnvNix {
     /// See the individual realisation functions for more details.
     // todo: return actual store paths built,
     // necessary when building manifest builds.
-    fn realise_lockfile(&self, lockfile: &Lockfile, system: &System) -> Result<(), BuildEnvError> {
+    fn realise_lockfile(
+        &self,
+        client: &impl ClientTrait,
+        lockfile: &Lockfile,
+        system: &System,
+    ) -> Result<(), BuildEnvError> {
         for package in lockfile.packages.iter() {
             if package.system() != system {
                 continue;
             }
             match package {
-                LockedPackage::Catalog(locked) => self.realise_nixpkgs(locked)?,
+                LockedPackage::Catalog(locked) => self.realise_nixpkgs(client, locked)?,
                 LockedPackage::Flake(locked) => self.realise_flakes(locked)?,
                 LockedPackage::StorePath(locked) => self.realise_store_path(locked)?,
             }
         }
         Ok(())
     }
 
+    /// Try to realise a custom package by asking for and downloading from the
+    /// recieved store info from the catalog.  Returns true if all outputs were
+    /// found and downloaedd, false otherwise.
+    fn try_realise_custom_pkg(
+        client: &impl ClientTrait,
+        locked: &LockedPackageCatalog,
+    ) -> Result<bool, BuildEnvError> {
+        // TODO - The API call accepts multiple, so an optimization is to
+        // collect these for the whole lockfile ahead of time and ask for them
+        // all at once.
+        let paths: Vec<String> = locked.outputs.values().map(|s| s.to_string()).collect();
+        let _store_locations = client
+            .get_store_info(paths)
+            .block_on()
+            .map_err(BuildEnvError::CatalogError)?;
+
+        // Try downloading each output from the store location provided.  If we
+        // are missing store info for any, we should return false.
+        // TODO - It is possible not all are missing. Do we need to check them
+        // each locally before attempting to download?  For now, we'll just
+        // assume they are all missing, noting custom packages only have one
+        // output currently.
```

Comment:
> It is possible not all are missing. Do we need to check them each locally before attempting to download?

`nix copy` will check and and only download if the path isnt present locally, returning with success immediately.


**... is what i would say** if nix would actually do that...
Unfortunately, nix seems to eagerly connect to the remote store.
apparently, even after connecting to a store nix doesnt check if the path is valid locally.


```
# custom nix build used by flox
nix path-info --offline /nix/store/7wr913d7zdn899ggl9p48gas243znipa-nix-2.18.8
/nix/store/7wr913d7zdn899ggl9p48gas243znipa-nix-2.18.8

# try copy it from a bad store
nix copy --from ssh://nope --to daemon /nix/store/7wr913d7zdn899ggl9p48gas243znipa-nix-2.18.8
ssh: Could not resolve hostname nope: nodename nor servname provided, or not known

# try copy from a good store
nix copy --from https://cache.nixos.org --to daemon /nix/store/7wr913d7zdn899ggl9p48gas243znipa-nix-2.18.8 
error: path '/nix/store/7wr913d7zdn899ggl9p48gas243znipa-nix-2.18.8' is required, but there is no substituter that can build it
```

hrm..

The quick solution would be `if self.check_store_path([path]) { continue; }`.

The more elaborate one is to extend `check_store_path` to tell you which paths are/arent valid and use that information to filter out the paths here.

I believe that what we eventually need anyway for #2489, but for now we can keep it at the simpler version.

Reply:
I think that as it's currently written, that check will have happened already in `realise_nixpkgs` when we get here.  So I'm leaning on leaving it be for now until we identify custom packages at the lockfile level and have a new match in `realise_lockfile`.  THEN it will need to be seaparate.

Reply:
Going to resolve based on this, but feel free to un-resolve with further comments.